src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs

```diff
@@ -405,13 +427,24 @@ protected IEnumerable<Symbol> GetUnsafeAddressTaken()
 
         #region Tracking reads/writes of variables for warnings
 
-        protected virtual void NoteRead(Symbol variable, ParameterSymbol rangeVariableUnderlyingParameter = null)
+        protected virtual void NoteRead(
+            Symbol variable,
+            ParameterSymbol rangeVariableUnderlyingParameter = null)
         {
             var local = variable as LocalSymbol;
             if ((object)local != null)
             {
-                _usedVariables.Add(local);
+                if (IsCapturedInLocalFunction(local,
+                                              rangeVariableUnderlyingParameter))
+                {
+                    NoteReadInLocalFunction(local, local);
```

Comment:
>NoteReadInLocalFunction [](start = 20, length = 23)

This appears to circumvent flow analysis of the local function, which is needed to determine if the variable is read before being written by the local function.